lib/membrane_rtmp_plugin/rtmp/source/tcp_server.ex

```diff
@@ -0,0 +1,35 @@
+defmodule Membrane.RTMP.Source.TcpServer do
+  @moduledoc """
+  A simple tcp server, which executes given function for each new connection.
+  """
+  use Task
+
+  @enforce_keys [:port, :tcp_options, :serve_fn]
+
+  defstruct @enforce_keys
+
+  @spec start_link(%__MODULE__{}) :: {:ok, pid}
+  def start_link(options) do
+    Task.start_link(__MODULE__, :run, [Enum.into(options, %{})])
+  end
+
+  @spec run(map()) :: nil
+  def run(options) do
+    {:ok, socket} = :gen_tcp.listen(options.port, options.tcp_options)
+
+    accept_loop(socket, options.serve_fn)
+  end
+
+  defp accept_loop(socket, serve_fn) do
+    {:ok, client} = :gen_tcp.accept(socket)
+    {:ok, pid} = Task.start(fn -> serve(client, serve_fn) end)
```

Comment:
I'm not sure if that logic is correct. The `serve` will spawn a process and try to give it a tcp control. It may happen that the `accept_loop` will do it later than the `serve` and you will have a race condition (serve can't give control unless itself being in control). You could spawn the serve task, give it control and then send it a message while `serve` will be waiting for the message and will continue with its logic.

Reply:
Yeah, you are right - race condition takes place here.